Feature adding category remaining crud operations

Amejia481 opened the pull request:
close #290

Amejia481 committed:
Adding remaining CRUD operations to categoryXXXLinks
#290

Adding delete and update operations to
CategoryCategoryComboLinkStore

Adding test for delete, update, delete table,
and query all to CategoryCategoryComboLinkStoreShould.

Adding delete and update operations to
CategoryCategoryOptionLinkStore

Adding delete and update operations to
and queryAll
CategoryOptionComboCategoryLinkStore

Adding test for delete, update, delete table,
and query all to CategoryCategoryComboLinkStoreShould.

Adding test for update, delete, delete table
and query all to CategoryOptionComboCategoryLinkStoreShould

Adding test for update, delete table
and query all to CategoryComboStoreShould

Renaming combo to categoryCombo
on CategoryCategoryComboLinkModel

Renaming option to categoryOption
on CategoryCategoryOptionLinkModel,
CategoryHandler

Renaming combo to categoryCombo and
option to categoryOption
on CategoryComboHandler

Renaming optionCombo to categoryOptionCombo
on CategoryOptionComboCategoryLinkModel

Adding CategoryComboMockFactory,CategoryMockFactory
and CategoryOptionMockFactory
to reduce duplicate code in tests

Add Override annotation by recommendation of checkstyle|Error prone
http://errorprone.info/bugpattern/MissingOverride

Fixing some warning of checkstyle|Error prone,
connectedAndroidTest

Amejia481 committed:
Merge remote-tracking branch 'origin/development' into feature-add_category_remaining_crud_operations

# Conflicts:
#	core/src/androidTest/java/org/hisp/dhis/android/core/category/CategoryCategoryComboLinkStoreShould.java

Amejia481 committed:
Adding test size annotation to new tests #290

xurxodev committed:
Merge branch 'development' into feature-add_category_remaining_crud_operations

xurxodev committed:
Remove SqliteCheckerUtility and use DatabaseAssert

xurxodev commented on core/src/main/java/org/hisp/dhis/android/core/category/CategoryOptionComboCategoryLinkStoreImpl.java:
```diff
@@ -37,14 +69,51 @@ public long insert(@NonNull CategoryOptionComboCategoryLinkModel entity) {
+    }
+
+    @Override
+    public boolean update(
```
To update one row only is necessary one model, you are pass as argument new and old, can you refactor this method?

Amejia481 replied:
Please, take a look at the example on CategoryCategoryOptionLink, the same applies here.

xurxodev commented on core/src/main/java/org/hisp/dhis/android/core/category/CategoryOptionComboCategoryLinkStoreImpl.java:
```diff
@@ -56,9 +125,58 @@ private int executeInsert() {
+    }
+
+    private List<CategoryOptionComboCategoryLinkModel>
+    mapCategoryOptionComboCategoryLinksModelsFromCursor(
```
this name is large, How only there is one method to map from cursor a name as mapFromCursor is more readable, can you rename it?

Amejia481 replied:
Sure!

xurxodev commented on core/src/main/java/org/hisp/dhis/android/core/category/CategoryCategoryComboLinkStoreImpl.java:
```diff
@@ -46,14 +64,48 @@ public long insert(CategoryCategoryComboLinkModel entity) {
+    }
+
+    @Override
+    public boolean update(@NonNull CategoryCategoryComboLinkModel oldCategoryCategoryComboLinkMode,
```
To update one row only is necessary one model, you are pass as argument new and old, can you refactor this method?

Amejia481 replied:
This case is similar to the one that I described you before all the link tables have two values and at the same time they are the primary key, with only one model, I'll lose the where clause.

xurxodev commented on core/src/main/java/org/hisp/dhis/android/core/category/CategoryCategoryOptionLinkStoreImpl.java:
```diff
@@ -4,47 +4,126 @@
+    }
+
+    @Override
+    public boolean update(
```
To update one row only is necessary one model, you are pass as argument new and old, can you refactor this method?

Amejia481 replied:
The problem with removing the extra parameter is that this table only has two columns(category and categoryOption) and at the same time they are the primary keys to this table for this reason, if I remove the second argument, I'll be forced to add two new parameters (category and categoryOption) as the where clause to identify which object, the client of the method wants to update, for example:

Imagine that I have this row on the table **CategoryCategoryOptionLink**

| category        | categoryOption |
| ------------- |:-------------:| 
| id_category_1      | id_categoryOption_1 |

And I want to update it to:

**CategoryCategoryOptionLink table**

| category        | categoryOption |
| ------------- |:-------------:| 
| id_category_2      | id_categoryOption_2 |

With the actual signature of the method I just pass:

**OLD (the where clause)**
``` 
 {
    "category": "category1", 
    "categoryOption":"categoryOption1"
}
```
**NEW (the data to update)**
``` 
 {
    "category": "category2", 
    "categoryOption":"categoryOption2"
}
``` 
With only one model, I'll lose the where clause.

xurxodev replied:
ok is true

xurxodev commented on core/src/main/java/org/hisp/dhis/android/core/category/CategoryCategoryComboLinkStoreImpl.java:
```diff
@@ -46,14 +64,48 @@ public long insert(CategoryCategoryComboLinkModel entity) {
+    public List<CategoryCategoryComboLink> queryAll() {
+        Cursor cursor = databaseAdapter.query(QUERY_ALL_CATEGORY_COMBO_LINKS);
+
+        return mapCategoryCategoryComboLinksFromCursor(cursor);
```
this name is large, How only there is one method to map from cursor a name as mapFromCursor is more readable, can you rename it?

Amejia481 replied:
Sure!

xurxodev commented on core/src/main/java/org/hisp/dhis/android/core/category/CategoryCategoryOptionLinkStoreImpl.java:
```diff
@@ -54,9 +133,39 @@ private int executeInsert() {
+        return wasExecuted(rowsAffected);
+    }
+
+    private List<CategoryCategoryOptionLinkModel> mapCategoryCategoryOptionLinkModelsFromCursor(
```
this name is large, How only there is one method to map from cursor a name as mapFromCursor is more readable, can you rename it?

Amejia481 replied:
Sure!

xurxodev commented on core/src/androidTest/java/org/hisp/dhis/android/core/common/CategoryComboMockFactory.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.ArrayList;
+import java.util.Date;
+
+public class CategoryComboMockFactory {
```
this class is not a factory, it's an [object mother](https://martinfowler.com/bliki/ObjectMother.html), can you refactor it?

Amejia481 replied:
Sure!

xurxodev commented on core/src/androidTest/java/org/hisp/dhis/android/core/common/CategoryMockFactory.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Date;
+
+public class CategoryMockFactory {
```
this class is not a factory, it's an [object mother](https://martinfowler.com/bliki/ObjectMother.html), can you refactor it?

xurxodev commented on core/src/androidTest/java/org/hisp/dhis/android/core/common/CategoryOptionMockFactory.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.util.Date;
+
+public final class CategoryOptionMockFactory {
```
this class is not a factory, it's an [object mother](https://martinfowler.com/bliki/ObjectMother.html), can you refactor it?

Amejia481 replied:
Sure!

xurxodev commented on core/src/androidTest/java/org/hisp/dhis/android/core/category/CategoryCategoryComboLinkStoreShould.java:
```diff
@@ -93,16 +161,45 @@ private void whenInsertNewCategoryCombo() {
     }
 
     private void whenInsertNewCategoryComboLink() {
```
this method has two responsibilities, create a fixture (givens...) and realize actions (whens..), I think that the code would be more readable in tests where is used if split this method in two, one responsibility per method, can you refactor it?

Amejia481 replied:
Sure!

xurxodev commented on core/src/androidTest/java/org/hisp/dhis/android/core/category/CategoryCategoryOptionLinkStoreShould.java:
```diff
@@ -34,52 +41,111 @@ public void setUp() throws IOException {
     }
 
     private void whenInsertNewCategoryOptionLink() {
+        givenACategory();
```
this method has two responsibilities, create a fixture (givens...) and realize actions (whens..), I think that the code would be more readable in tests where is used if split this method in two, one responsibility per method, can you refactor it?

Amejia481 replied:
Sure!

xurxodev commented on core/src/androidTest/java/org/hisp/dhis/android/core/category/CategoryComboStoreShould.java:
```diff
@@ -56,33 +58,60 @@ public void insert_and_delete_a_category_combo() throws Exception {
     }
 
     private void whenInsertNewCategoryCombo() {
+
```
this method has two responsibilities, create a fixture (givens...) and realize actions (whens..), I think that the code would be more readable in tests where is used if split this method in two, one responsibility per method, can you refactor it?

Amejia481 replied:
Sure!

xurxodev commented on core/src/androidTest/java/org/hisp/dhis/android/core/category/CategoryOptionComboCategoryLinkStoreShould.java:
```diff
@@ -33,55 +41,86 @@ public void setUp() throws IOException {
     }
 
     private void whenInsertNewCategoryOptionComboCategoryLink() {
+        givenACategoryOption();
```
this method has two responsibilities, create a fixture (givens...) and realize actions (whens..), I think that the code would be more readable in tests where is used if split this method in two, one responsibility per method, can you refactor it?

Amejia481 replied:
Sure!

xurxodev reviewed (changes requested):
@Amejia481 I have requested you some changes, can you review it, please?

Amejia481 committed:
Fixing DataBaseMigrationShould tests

org.hisp.dhis.android.core.data.database.migrations.DataBaseMigrationShould#not_have_category_table_after_downgrade_with_database_version_2
org.hisp.dhis.android.core.data.database.migrations.DataBaseMigrationShould#have_dropped_table_when_down_migration_drop_table
org.hisp.dhis.android.core.data.database.migrations.DataBaseMigrationShould#not_have_category_table_after_downgrade_with_real_sql_database

Amejia481 committed:
Refactoring tests

xurxodev commented on core/src/main/java/org/hisp/dhis/android/core/category/CategoryCategoryComboLinkStoreImpl.java:
```diff
@@ -46,14 +64,48 @@ public long insert(CategoryCategoryComboLinkModel entity) {
+    }
+
+    @Override
+    public boolean delete(@NonNull CategoryCategoryComboLinkModel entity) {
```
In all stores in this method that delete by uid returns a int type, please can you refactor it?

Amejia481 replied:
I changed the return type to an int but this entity is a link table and it has two primary keys if I remove the object I will have to have two uids instead of one object.

Do you have that I use two uids instead of one object?

xurxodev commented on core/src/main/java/org/hisp/dhis/android/core/category/CategoryOptionComboCategoryLinkStoreImpl.java:
```diff
@@ -37,14 +69,51 @@ public long insert(@NonNull CategoryOptionComboCategoryLinkModel entity) {
+    }
+
+    @Override
+    public boolean delete(@NonNull CategoryOptionComboCategoryLinkModel element) {
```
In all stores in this method that delete by uid returns a int type, please can you refactor it?

Amejia481 replied:
I changed the return type to an int but this entity is a link table and it has two primary keys if I remove the object I will have to have two uids instead of one object.

Do you have that I use two uids instead of one object?

xurxodev commented on core/src/main/java/org/hisp/dhis/android/core/category/CategoryCategoryOptionLinkStoreImpl.java:
```diff
@@ -4,47 +4,126 @@
+    }
+
+    @Override
+    public boolean delete(@NonNull CategoryCategoryOptionLinkModel categoryOptionLinkModel) {
```
In all stores in this method that delete by uid returns a int type, please can you refactor it?

Amejia481 replied:
I changed the return type to an int but this entity is a link table and it has two primary keys if I remove the object I will have to have two uids instead of one object.

Do you have that I use two uids instead of one object?

xurxodev reviewed (changes requested):
@Amejia481 I have found some little changes more, can you review it, please?

Amejia481 committed:
Refactoring delete methods on the category stores

xurxodev reviewed (changes requested):
please, @Amejia481 Can you refactor also update methods to return int value as in all stores? thanks

Amejia481 committed:
Refactoring update signature methods of category stores

xurxodev committed:
Merge branch 'development' into feature-add_category_remaining_crud_operations

xurxodev reviewed (approved):
thanks @Amejia481
